aggregator/src/core.rs

```diff
@@ -62,3 +82,674 @@ pub(crate) fn extract_accumulators_and_proof(
         )?;
     Ok((accumulator, transcript_write.finalize()))
 }
+
+/// Input the hash input bytes,
+/// assign the circuit for the hash function,
+/// return
+/// - cells of the hash digests
+/// - the cell that contains the number of valid snarks
+//
+// This function asserts the following constraints on the hashes
+//
+// 1. batch_data_hash digest is reused for public input hash
+// 2. batch_pi_hash used same roots as chunk_pi_hash
+// 2.1. batch_pi_hash and chunk[0] use a same prev_state_root
+// 2.2. batch_pi_hash and chunk[MAX_AGG_SNARKS-1] use a same post_state_root
+// 2.3. batch_pi_hash and chunk[MAX_AGG_SNARKS-1] use a same withdraw_root
+// 3. batch_data_hash and chunk[i].pi_hash use a same chunk[i].data_hash when chunk[i] is not padded
+// 4. chunks are continuous: they are linked via the state roots
+// 5. batch and all its chunks use a same chain id
+// 6. chunk[i]'s prev_state_root == post_state_root when chunk[i] is padded
+// 7. chunk[i]'s data_hash == "" when chunk[i] is padded
+#[allow(clippy::type_complexity)]
+pub(crate) fn assign_batch_hashes(
+    config: &AggregationConfig,
+    layouter: &mut impl Layouter<Fr>,
+    challenges: Challenges<Value<Fr>>,
+    preimages: &[Vec<u8>],
+    num_of_valid_chunks: usize,
+) -> Result<(Vec<AssignedCell<Fr, Fr>>, AssignedCell<Fr, Fr>), Error> {
+    let (hash_input_cells, hash_output_cells, data_rlc_cells, hash_input_len_cells) =
+        extract_hash_cells(
+            &config.keccak_circuit_config,
+            layouter,
+            challenges,
+            preimages,
+        )?;
```

Comment:
spend some time to think about this. 

> The cells in aggregation circuit are too tightly coupled with keccak sub-circuit. It's hard to audit the code.

is there a case where we modify the keccak circuit while not require changes to this pi aggregation circuit? I think not.

So the real complaint here is that the code is hard coded for `MAX_AGG_SNARK`. This I agree.

OTOH I think the keccak circuit itself is not well designed. I.e., currently, it uses 87 columns and 2^19 rows, i.e., roughly 43 million cells, supporting 1600ish keccak rounds; while we actually only used 25 rounds, or, 87 * 25 * 300 = 652 K cells. If we stick with k=20, then we should be able to fill all those in a single column or two, rather than 87 columns. This will result into 3x in the overall aggregator performance. 

(there are also quite a few wasted cells within those 652k cells as per currently layout of the keccak table)

I think a refactor of (semi-static) keccak circuit is of a higher priority than making the above code less hard-coded.

Reply:
> is there a case where we modify the keccak circuit while not require changes to this pi aggregation circuit? I think not.

But the design goal is this aggregation circuit should not need any change even if we change the layout of keccak circuit to a more thinner one, right?
For now, if we shrink the keccak circuit's layout from 87 cols to like 29 cols (num_rows_per_round: 12 -> 50), then we need at least refactor the `get_indices` function which is an example of tight coupling that I worried.